### PR TITLE
MCI API Support

### PR DESCRIPTION
**Pull request by yakloinsteak**

### CODE REVIEW

File: drivers/hmis_external_apis/README.md
Comment (eanders) [inline drivers/hmis_external_apis/README.md]: Can you add a brief overview of what this driver does here?

Comment (yakloinsteak) [reply drivers/hmis_external_apis/README.md]: I was planning to do that after I know more about what it's supposed to do.  I'm a little blocked on specs for this branch.

File: drivers/hmis_external_apis/app/models/hmis_external_apis/mci_payload.rb
Comment (gigxz) [inline drivers/hmis_external_apis/app/models/hmis_external_apis/mci_payload.rb]: `birthDate` is required for `ClientInsertVM` and `ClientUpdateVM` payload schemas. So if that is the date you're having issues with, that's why. We should error if any of `client.first_name` `client.last_name` or `client.dob` are missing.

File: drivers/hmis_external_apis/app/models/hmis_external_apis/mci_payload.rb
Comment (gigxz) [inline drivers/hmis_external_apis/app/models/hmis_external_apis/mci_payload.rb]: @yakloinsteak this looks great. I can fill in the mappings for race, gender, and ethnicity after this is merged in.
```suggestion
      payload = {
        # 'userID' => 'string',
        'otherNames' => {
          'firstName' => client.preferred_name,
          'lastName' => client.preferred_name,
        },
        # 'isHomeless' => true,
        'ssnAlias' => client.ssn, # Should be last 4 digits of ssn, if we only have that
        'ethnicityCode' => 7, # Gig to do mapping
        'ethnicityDesc' => 'WCDesc', # I would guess not needed, but can pull from ::HudLists.ethnicity_map
        # 'tribeCode' => 414,
        # 'tribeDesc' => 'WCDesc',
        # 'maritalStatus' => 8,
        # 'housingStatus' => 15,
        'genderCode' => 1, # Gig to do mapping
        # 'residencyCode' => 3,
        'firstName' => client.first_name,
        'middleName' => client.middle_name,
        'lastName' => client.last_name,
        'suffix' => client.name_suffix,
        'ssn' => client.ssn, # Should be complete SSN, if we have it
        'birthDate' => client.dob.to_s(:db),
        'raceCodes' => '2-,', # Gig to do mapping
        # 'deathDate' => '2019-05-09T11:23:31.129Z',
      }
```

File: drivers/hmis_external_apis/app/models/hmis_external_apis/mci_payload.rb
Comment (gigxz) [inline drivers/hmis_external_apis/app/models/hmis_external_apis/mci_payload.rb]: First, last, middle, suffix, dob, and SSN would be a good starting point. I can add in the reverse race/ethnicity/gender mappings after.

File: db/warehouse/migrate/20230406183420_client_mci_id.rb
Comment (gigxz) [inline db/warehouse/migrate/20230406183420_client_mci_id.rb]: Based on our last conversation:
```suggestion
    #  t.string :value
    #  t.references :source, polymorphic: true
    #  t.references :remote_credential
    #  references :external_request_log
    #  t.timestamps
```

Theron's PR https://github.com/greenriver/hmis-warehouse/pull/2955 requires this table and the external driver to exist. I'm happy to merge this scaffolding in when it's ready @yakloinsteak and the remaining API bits can be finished up separately.

File: drivers/hmis_external_apis/app/models/hmis_external_apis/mci_payload.rb
Comment (yakloinsteak) [reply drivers/hmis_external_apis/app/models/hmis_external_apis/mci_payload.rb]: It was more a general problem as I reversed engineered the API. We don't get a "birthDate was invalid" error but instead useless errors that don't help know the problem.

**Comment by yakloinsteak** [comment]
There were more commits that I hadn't pushed yet (I'm not used to code reviews while PRs are still in draft state). I'll be more careful moving forward so I don't waste @eanders or @gigxz 's time. Sorry about that.

**Comment by gigxz** [comment]
Oh that's alright! In the future I will hold off until a review is requested

File: db/warehouse/migrate/20230406183420_client_mci_id.rb
Comment (yakloinsteak) [reply db/warehouse/migrate/20230406183420_client_mci_id.rb]: I'm going to resolve this since I hadn't pushed my latest work yet. Sorry again about that.

**Comment by eanders** [comment]
@yakloinsteak, same here, I don't usually do draft-PR review, but sometimes (especially on new code) it's nice to see where things are heading and drop a few notes in there to head off any early issues.  I'll hold off on future PRs as well, unless you ask.

**Comment by yakloinsteak** [comment]
I'm giving up on trying to load client data and doing further testing for now. I can't get minio to work yet (so I can't load client data). If somebody could help, I can proceed with more testing. The problem has to do with "The request signature we calculated does not match the signature you provided. Check your key and signing method." I suppose I could just use my own bucket, but I imagine this is a solved problem. I'm happy to keep at this if somebody can help me load some data.

File: drivers/hmis_external_apis/app/models/hmis_external_apis/mci.rb
Comment (yakloinsteak) [inline drivers/hmis_external_apis/app/models/hmis_external_apis/mci.rb:7]: @gigxz This is the only step you need before using the API in this class.

File: db/warehouse/migrate/20230406183420_client_mci_id.rb
Comment (gigxz) [inline db/warehouse/migrate/20230406183420_client_mci_id.rb]: I think this should include remote_credential too. Technically if we have 2 external systems, a client could conceivably have the same external ID on each.

File: drivers/hmis_external_apis/app/models/hmis_external_apis/mci.rb
Comment (gigxz) [inline drivers/hmis_external_apis/app/models/hmis_external_apis/mci.rb]: Did I do that right?

File: db/warehouse/migrate/20230406183420_client_mci_id.rb
Comment (yakloinsteak) [reply db/warehouse/migrate/20230406183420_client_mci_id.rb]: That makes sense. I'll update it.

File: drivers/hmis_external_apis/app/models/hmis_external_apis.rb
Comment (yakloinsteak) [inline drivers/hmis_external_apis/app/models/hmis_external_apis.rb:4]: @gigxz What's your take on table prefixes? I know there are a ton of tables in this app, so maybe we should have one.

File: drivers/hmis_external_apis/app/models/hmis_external_apis/mci.rb
Comment (yakloinsteak) [reply drivers/hmis_external_apis/app/models/hmis_external_apis/mci.rb]: The url would be `result.request` and request would be `payload`, but that's confusing. I have data now (yippie!) so I can iterate quicker now. I'll fix the wording of the fields to make this clearer. Thanks for bringing it up.

File: drivers/hmis_external_apis/app/models/hmis_external_apis.rb
Comment (gigxz) [reply drivers/hmis_external_apis/app/models/hmis_external_apis.rb:4]: I don't have a take, I'm not really sure how this works in rails. The tables are already prefixed with `external_`, and the models are happy, so I think we can leave this out..
@eanders?

File: drivers/hmis_external_apis/app/models/hmis_external_apis/external_id.rb
Comment (gigxz) [inline drivers/hmis_external_apis/app/models/hmis_external_apis/external_id.rb]: It will be references outside the driver, but that's okay. The hmis driver has an implicit "dependency" on this driver.

File: drivers/hmis_external_apis/app/models/hmis_external_apis.rb
Comment (eanders) [reply drivers/hmis_external_apis/app/models/hmis_external_apis.rb:4]: Yeah, I've mostly been leaving out the table name prefixes.  I think it's fine to just go with `external_` for these.  If you want to make that an explicit prefix rather than encoded in each model, that's fine, but not necessary.

File: drivers/hmis_external_apis/app/models/hmis_external_apis/mci.rb
Comment (gigxz) [inline drivers/hmis_external_apis/app/models/hmis_external_apis/mci.rb]: This can be removed as mentioned on PT. Keep `build_client`, we will need it for referrals

**Review by gigxz — APPROVED** [review]
This looks good to me to merge, just to get the infra in & unblock related work!

Remaining work on this code:
* Error handling (Todd)
* Race/gender mappings (Gig)

**Comment by yakloinsteak** [comment]
@gigxz Let me know if you want another look, but I plan to merge this after the tests pass. I'll ignore the brakeman issues unless you want me to deal with that (I'm happy to do that).

**Comment by eanders** [comment]
@yakloinsteak if you have time to bump nokogiri, that'd be great.

**Comment by gigxz** [comment]
Error and logging updates look good to me! Thanks!

File: db/warehouse/migrate/20230406183420_client_mci_id.rb
Comment (ttoomey) [inline db/warehouse/migrate/20230406183420_client_mci_id.rb:33]: @yakloinsteak 
If I understand correctly, it looks like this would allow duplicate external_ids if they reference different sources. For example, two client records (sources) could have external_id records that have the same remote_credential/ value.

I imagine we shouldn't allow that? Maybe we could have one constraint for remote_credential and one for source, something like this?
```ruby
     # a source can only have one external_id for a given remote_credential
      t.index [:source_id, :source_type, :remote_credential_id], unique: true, name: "external_ids_uniq_1"
     # remote_credential ids must be unique for a given source_type
      t.index [:source_type, :remote_credential_id, :value], unique: true, name: "external_ids_uniq_2"
```

Comment (yakloinsteak) [reply db/warehouse/migrate/20230406183420_client_mci_id.rb:33]: Feel free to make those changes if @gigxz agrees. Maybe a remote system could have multiple IDs for a source? I don't know the landscape well enough, but I think what you're saying makes sense. Definitely :+1: for more constraints when you can.

Comment (eanders) [reply db/warehouse/migrate/20230406183420_client_mci_id.rb:33]: I believe the MCI might have more than one ID per person.  That was unclear to me.  I believe **remote_credential ids must be unique for a given source_type** is a hard-fast rule.

Comment (gigxz) [reply db/warehouse/migrate/20230406183420_client_mci_id.rb:33]: No, that's not quite right. At least for MCI:
1. One Client in HMIS has at most 1 MCI ID
2. Two Clients in HMIS can share the same MCI ID

Given (2), we do not want the constraint "remote_credential ids must be unique for a given source_type".

Comment (eanders) [reply db/warehouse/migrate/20230406183420_client_mci_id.rb:33]: Hmm, ok, I thought there were situations where someone was merged but the two clients had been previously given different MCI IDs so they may have more than one...

Comment (gigxz) [reply db/warehouse/migrate/20230406183420_client_mci_id.rb:33]: They confirmed the above at Mondays meeting. But maybe we should ask again in the context of merging clients specifically. We haven't had an in depth conversation around merging yet.

Comment (ttoomey) [reply db/warehouse/migrate/20230406183420_client_mci_id.rb:33]: > Two Clients in HMIS can share the same MCI ID

@gigxz I see- so if we receive a referral request for a household member and need to match to an existing client using the MCI ID, we could find more than one matching record. I guess in that case we could try and match on secondary fields (name, social, dob). And then we'd have to use ID order as a fallback.

Comment (gigxz) [reply db/warehouse/migrate/20230406183420_client_mci_id.rb:33]: Yeah, that's what I was thinking, I'm planning to confirm with them about this. There's an open question on whether we should actually _update_ our demographic fields based on the data we receive via a referral request.

Comment (gigxz) [reply db/warehouse/migrate/20230406183420_client_mci_id.rb:33]: Don't worry about it for now, you can just take the first matching client.

**Pull request merged**
